core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -51,29 +63,66 @@ package object config {
     ConfigBuilder(SparkLauncher.EXECUTOR_EXTRA_LIBRARY_PATH).stringConf.createOptional
 
   private[spark] val EXECUTOR_USER_CLASS_PATH_FIRST =
-    ConfigBuilder("spark.executor.userClassPathFirst").booleanConf.createWithDefault(false)
+    ConfigBuilder("spark.executor.userClassPathFirst")
+      .doc("Same functionality as spark.driver.userClassPathFirst, " +
+        "but applied to executor instances.")
+      .booleanConf
+      .checkValues(Set(false, true))
+      .createWithDefault(false)
 
   private[spark] val EXECUTOR_MEMORY = ConfigBuilder("spark.executor.memory")
+    .doc("Amount of memory to use per executor process (e.g. 2g, 8g).")
     .bytesConf(ByteUnit.MiB)
+    .checkValue(v => v > 0 && v <= Int.MaxValue / (1024 * 1024),
+      "The executor memory must be greater than 0M " +
+      s"and less than ${Int.MaxValue / (1024 * 1024)}M.")
     .createWithDefaultString("1g")
 
-  private[spark] val IS_PYTHON_APP = ConfigBuilder("spark.yarn.isPython").internal()
-    .booleanConf.createWithDefault(false)
+  private[spark] val IS_PYTHON_APP = ConfigBuilder("spark.yarn.isPython")
+    .internal()
+    .booleanConf
+    .checkValues(Set(false, true))
+    .createWithDefault(false)
 
-  private[spark] val CPUS_PER_TASK = ConfigBuilder("spark.task.cpus").intConf.createWithDefault(1)
+  private[spark] val CPUS_PER_TASK = ConfigBuilder("spark.task.cpus")
+    .doc("Number of cores to allocate for each task. ")
+    .intConf
+    .checkValue(v => v > 0,
+      "Number of cores to allocate for task event queue must not be negative")
+    .createWithDefault(1)
 
   private[spark] val DYN_ALLOCATION_MIN_EXECUTORS =
-    ConfigBuilder("spark.dynamicAllocation.minExecutors").intConf.createWithDefault(0)
+    ConfigBuilder("spark.dynamicAllocation.minExecutors")
+      .doc("Lower bound for the number of executors if dynamic allocation is enabled.")
+      .intConf
+      .checkValue(v => v >= 0,
+        "Lower bound for the number of executors should be greater than or equal to 0")
+      .createWithDefault(0)
 
   private[spark] val DYN_ALLOCATION_INITIAL_EXECUTORS =
     ConfigBuilder("spark.dynamicAllocation.initialExecutors")
+      .doc("Initial number of executors to run if dynamic allocation is enabled. " +
+        "If `--num-executors` (or `spark.executor.instances`) is set " +
+        "and larger than this value, " +
+        "it will be used as the initial number of executors.")
       .fallbackConf(DYN_ALLOCATION_MIN_EXECUTORS)
 
   private[spark] val DYN_ALLOCATION_MAX_EXECUTORS =
-    ConfigBuilder("spark.dynamicAllocation.maxExecutors").intConf.createWithDefault(Int.MaxValue)
+    ConfigBuilder("spark.dynamicAllocation.maxExecutors")
+      .doc("Upper bound for the number of executors if dynamic allocation is enabled. ")
+      .intConf
+      .createWithDefault(Int.MaxValue)
```

Comment:
It might be a bit out of place.
In org.apache.spark.ExecutorAllocationManagerSuite.scala
`val conf2 = conf.clone().set("spark.dynamicAllocation.maxExecutors", "-1")`
Why is this value set less than zero?
Not yet understood.